pkgs/development/interpreters/python/python-packages-base.nix

```diff
@@ -14,24 +14,31 @@ let
 
   # Derivations built with `buildPythonPackage` can already be overridden with `override`, `overrideAttrs`, and `overrideDerivation`.
   # This function introduces `overridePythonAttrs` and it overrides the call to `buildPythonPackage`.
-  makeOverridablePythonPackage = f: origArgs:
+  makeOverridablePythonPackage = f: lib.copyFunctionArgs f (origArgs:
     let
-      args = lib.fix (lib.extends
+      # If rattrs is a function, apply it to attrs, or else return rattrs.
+      applyRattrs = rattrs: attrs: if lib.isFunction rattrs then rattrs attrs else rattrs;
+
+      argsRecursive = lib.extends
         (_: previousAttrs: {
           passthru = (previousAttrs.passthru or { }) // {
             overridePythonAttrs = newArgs: makeOverridablePythonPackage f (overrideWith newArgs);
           };
         })
-        (_: origArgs));
+        (if lib.isFunction origArgs then origArgs else (_: origArgs));
+
+      args = if lib.isFunction origArgs then argsRecursive else lib.fix argsRecursive;
+
       result = f args;
-      overrideWith = newArgs: args // (if pkgs.lib.isFunction newArgs then newArgs args else newArgs);
+      overrideWithRecursive = newArgs: finalAttrs: argsRecursive finalAttrs // applyRattrs newArgs (argsRecursive finalAttrs);
+      overrideWith = if lib.isFunction origArgs then overrideWithRecursive else lib.fix (lib.flip overrideWithRecursive);
     in
       if builtins.isAttrs result then result
       else if builtins.isFunction result then {
         overridePythonAttrs = newArgs: makeOverridablePythonPackage f (overrideWith newArgs);
         __functor = self: result;
       }
-      else result;
+      else result);
```

Comment:
This feels too custom or repeated (not sure yet which). Does this add `mkDerivation`-like fixpoint logic at the python attrs level?

I believe we should merge the python attrs level into the mkDerivation attrs, so that the interface and implementation become simpler. Having multiple levels of overriding has a huge complexity cost, so getting rid of an unnecessary level would be a huge win. We'd get rid of `overridePythonAttrs` and all the user facing complexity, implementation complexity and bugs that come with it.

The python-specific attrs can *almost* be implemented as an "overlay" on the `mkDerivation` arguments. When I tried this, I think only like 3 attributes had the same name but a slightly different meaning. That made it a breaking change, but migrating those attributes is feasible and would vastly simplify the python/mkDerivation wiring.

Reply:
With same names that is generally speaking because the Python builder will extend the lists with some "defaults". That's really the only value of the custom builder over just plain hooks.

Reply:
> This feels too custom or repeated (not sure yet which). Does this add `mkDerivation`-like fixpoint logic at the python attrs level?

I personally prefer passing most of the attributes into `mkDerivation`, encourage the use of `overrideAttrs` and gradually deprecates `overridePythonAttrs` as well as other builder-specific override methods. That will be a mass rebuild, so I just work around the `makeOverridablePythonPackage` obstacle in order to demonstrate the possibility to add the recursive attributes support without rebuild.